### PR TITLE
Introduce new Constant

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
This PR is dependent on all of COFFEE being removed first.

## Associated Pull Requests:
- [701](https://github.com/OP2/PyOP2/pull/701)
- [294](https://github.com/firedrakeproject/tsfc/pull/294)
- [98](https://github.com/dolfin-adjoint/pyadjoint/pull/98)
- After Firedrake is merged this can go in also: [51](https://bitbucket.org/pefarrell/defcon/branch/JDBetteridge/enable_arclength_continuation)

## Fixes Issues:
- Maybe

# Checklist for author:

- [x] I have linted the codebase (`make lint` in the `firedrake` source directory).
- [x] My changes generate no new warnings.
- [x] All of my functions and classes have appropriate docstrings.
- [x] I have commented my code where its purpose may be unclear.
- [x] I have included and updated any relevant documentation.
- [x] Documentation builds locally (`make linkcheck; make html; make latexpdf` in `firedrake/docs` directory)
- [x] I have added tests specific to the issues fixed in this PR.
- [x] I have added tests that exercise the new functionality I have introduced
- [x] Tests pass locally (`pytest tests` in the `firedrake` source directory) (useful, but not essential if you don't have suitable hardware).
- [x] I have performed a self-review of my own code using the below guidelines.

# Checklist for reviewer:

- [ ] Docstrings present
- [ ] New tests present
- [ ] Code correctly commented
- [ ] No bad "code smells"
- [ ] No issues in parallel
- [ ] No CI issues (excessive parallelism/memory usage/time/warnings generated)
- [ ] Upstream/dependent branches and PRs are ready

### CODE REVIEW

File: firedrake/constant.py [hunk 1]
```diff
@@ -14,19 +15,24 @@
...
     else:
-        dat = op2.Global(shape, data, comm=comm)
+        dat = op2type(shape, data, comm=comm)
     return dat, rank, shape
 
 
-class Constant(ufl.Coefficient, ConstantMixin):
-
+# Think "literal"
+class Constant(ufl.constantvalue.ConstantValue, ConstantMixin):
```
Comment (wence-) [inline firedrake/constant.py:35]: Surely the right thing to inherit from is https://github.com/FEniCS/ufl/blob/d2dfbef9a60bede67ac8361712316bf427f9eecc/ufl/constant.py#L18

?

`ConstantValue` is a compile-time literal in UFL, not a runtime constant.

Comment (connorjward) [reply firedrake/constant.py:35]: The reason we have done this is that a UFL `Constant` is what Firedrake calls a `Function` in R. This means that it has an associated `ufl_domain` and therefore a communicator. The reason for the changes being made here is that many people use `Constant` just for parametrising kernel arguments and in some cases (like defcon) finding an appropriate mesh/communicator is tricky and we get deadlocks.

Our solution, not quite shown here, is for TSFC to define a `KernelArgument` mixin class that marks an element of a UFL expression/form as being a kernel parameter. Firedrake `Function` and `Constant` would then inherit from this.

Comment (wence-) [reply firedrake/constant.py:35]: This conversely means that `Constant(1)*dx` needs one to define a domain on `dx`.

Comment (wence-) [reply firedrake/constant.py:35]: > The reason we have done this is that a UFL Constant is what Firedrake calls a Function in R.

This is not quite true because you can't have a `Constant` as an unknown (for a solution variable, etc..)

File: firedrake/assemble.py [hunk 2]
```diff
@@ -1229,6 +1244,16 @@ def iter_active_coefficients(form, kinfo):
             for subidx in subidxs:
                 yield form.coefficients()[idx].subfunctions[subidx]
 
+    @staticmethod
+    def iter_constants(form, kinfo):  # FIXME kinfo not currently needed
```
[the remark this answers is not in the document]
Comment (ReubenHill) [reply firedrake/assemble.py:1248]: I guess from lookign at the rest of this file, it seems that if a function's name is descriptive enough then it doesn't need a docstring

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/ufl_expr.py [hunk 3]
```diff
@@ -155,43 +158,48 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
...
     elif isinstance(uc, firedrake.Constant):
         if uc.ufl_shape != ():
             raise ValueError("Real function space of vector elements not supported")
+        # Replace instances of the constant with a new argument ``x``
+        # and differentiate wrt ``x``.
         V = firedrake.FunctionSpace(mesh, "Real", 0)
-        du = argument(V)
+        x = ufl.Coefficient(V, n + 1)
+        n += 1
+        form = ufl.replace(form, {u: x})
```
[the remark this answers is not in the document]
Comment (JDBetteridge) [reply firedrake/ufl_expr.py:183]: https://github.com/FEniCS/ufl/issues/171

File: tests/regression/test_solver_error_checking.py [hunk 4]
```diff
@@ -37,6 +37,9 @@ def test_invalid_solution_type(a, L, c):
         solve(a == L, c)
 
 
+# TODO: Check with a Constant with a domain (ie: add check for vector lengths)
```
[the remark this answers is not in the document]
Comment (JDBetteridge) [reply tests/regression/test_solver_error_checking.py:40]: This comment is only relevant to a previous state of the PR. Function in the R-space (even if it's constructed as a Constant with a domain) is a valid solution type.

File: firedrake/adjoint/variational_solver.py [hunk 5]
```diff
@@ -105,6 +104,7 @@ def _ad_problem_clone(self, problem, dependencies):
         expressions, we'll instead create clones of them.
         """
         from firedrake import NonlinearVariationalProblem
+        from firedrake.function import Function
```
Comment (connorjward) [inline firedrake/adjoint/variational_solver.py:107]:
```suggestion
        from firedrake import Function, NonlinearVariationalProblem
```

File: firedrake/constant.py [hunk 6]
```diff
@@ -49,43 +53,56 @@ class Constant(ufl.Coefficient, ConstantMixin):
...
+    def __new__(cls, value, domain=None):
+        if domain:
+            # Avoid circular import
+            from firedrake.function import Function
+            from firedrake.functionspace import FunctionSpace
+            import warnings
+            warnings.warn("Constants with a domain defined are functions in the Real space")
+
+            dat, rank, shape = _create_dat(op2.Global, value, domain._comm)
+            dat.zero()
```
Comment (connorjward) [inline firedrake/constant.py:67]: Why are we zeroing here? We have a defined `value`.

File: firedrake/constant.py [hunk 7]
```diff
@@ -49,43 +53,56 @@ class Constant(ufl.Coefficient, ConstantMixin):
...
     """
+    _globalcount = 0
+
+    def __new__(cls, value, domain=None):
+        if domain:
+            # Avoid circular import
+            from firedrake.function import Function
+            from firedrake.functionspace import FunctionSpace
+            import warnings
+            warnings.warn("Constants with a domain defined are functions in the Real space")
```
Comment (connorjward) [inline firedrake/constant.py:64]:
```suggestion
            warnings.warn(
                "Giving Constants a domain is no longer supported. Instead please "
                "create a Function in the Real space.", FutureWarning
            )
```

File: firedrake/constant.py [hunk 8]
```diff
@@ -49,43 +53,56 @@ class Constant(ufl.Coefficient, ConstantMixin):
...
+            cell = domain.ufl_cell()
+            if rank == 0:
+                element = ufl.FiniteElement("R", cell, 0)
+            elif rank == 1:
+                element = ufl.VectorElement("R", cell, 0, shape[0])
+            else:
+                element = ufl.TensorElement("R", cell, 0, shape=shape)
+
+            R = FunctionSpace(domain, element, name="firedrake.Constant")
+            return Function(R, val=dat).assign(value)
```
Comment (connorjward) [inline firedrake/constant.py:79]: Same question about `value`.

File: firedrake/function.py [hunk 9]
```diff
@@ -413,6 +415,14 @@ def assign(self, expr, subset=None):
         """
         if expr == 0:
             self.dat.zero(subset=subset)
+        # Assign literal collections to vector and tensor valued functions in R.
+        elif (isinstance(expr, collections.abc.Container)
+              and self.ufl_element().family() == "Real"):
```
Comment (connorjward) [inline firedrake/function.py:420]:
```suggestion
        elif self.ufl_element().family() == "Real":
```

I think the extra `isinstance` check is overkill. Also won't `f.assign(non_vector_value)` fail with it when we would expect it to work?

File: firedrake/projection.py [hunk 10]
```diff
@@ -282,7 +282,8 @@ def Projector(v, v_out, bcs=None, solver_parameters=None,
     else:
         if bcs is not None:
             raise ValueError("Haven't implemented supermesh projection with boundary conditions yet, sorry!")
-        if not isinstance(source, function.Function):
+        # Test if not a function OR whether unknown in the reals
```
Comment (connorjward) [inline firedrake/projection.py:285]:
```suggestion
```

Not a useful comment.

File: firedrake/slate/slac/kernel_builder.py [hunk 11]
```diff
@@ -107,13 +108,16 @@ def shape(self, tensor):
         else:
             return tensor.shape
 
-    def extent(self, coefficient):
-        """ Calculation of the range of a coefficient."""
-        element = coefficient.ufl_element()
-        if element.family() == "Real":
-            return (coefficient.dat.cdim, )
+    def extent(self, argument):
+        """ Calculation of the value size of a constant or coefficient."""
```
Comment (connorjward) [inline firedrake/slate/slac/kernel_builder.py:112]:
```suggestion
        """Return the value size of a constant or coefficient."""
```

File: firedrake/solving.py [hunk 12]
```diff
@@ -141,6 +140,10 @@ def _solve_varproblem(*args, **kwargs):
         near_nullspace, \
         options_prefix = _extract_args(*args, **kwargs)
 
+    # Check whether solution is valid
+    if not isinstance(u, (function.Function, vector.Vector)):
+        raise TypeError("Provided solution is a '%s', not a Function" % type(u).__name__)
```
Comment (connorjward) [inline firedrake/solving.py:145]:
```suggestion
        raise TypeError(f"Provided solution is a '{type(u).__name__}', not a Function")
```

File: firedrake/solving.py [hunk 13]
```diff
@@ -218,6 +221,10 @@ def _la_solve(A, x, b, **kwargs):
     bcs, solver_parameters, nullspace, nullspace_T, near_nullspace, \
         options_prefix = _extract_linear_solver_args(A, x, b, **kwargs)
 
+    # Check whether solution is valid
+    if not isinstance(x, (function.Function, vector.Vector)):
+        raise TypeError("Provided solution is a '%s', not a Function" % type(x).__name__)
```
Comment (connorjward) [inline firedrake/solving.py:226]:
```suggestion
        raise TypeError(f"Provided solution is a '{type(x).__name__}', not a Function")
```

File: firedrake/tsfc_interface.py [hunk 14]
```diff
@@ -201,6 +201,7 @@ def compile_form(form, name, parameters=None, split=True, interface=None, diagon
 
     key = (name, utils.tuplify(parameters), split, diagonal)
     try:
+        raise KeyError
```
Comment (connorjward) [inline firedrake/tsfc_interface.py:204]:
```suggestion
```

This should definitely go!

**Review by connorjward — CHANGES_REQUESTED** [review]

File: firedrake/constant.py (hunk 8 above)
Comment (JDBetteridge) [reply firedrake/constant.py:82]: The assign needs to happen for annotation to occur. We can get away with not zeroing the `dat`, but we need to assign to create an assign block.

File: .github/workflows/build.yml [hunk 15]
```diff
@@ -74,7 +74,10 @@ jobs:
             --install gusto \
             --install icepack \
             --install irksome \
-            --install femlium || (cat firedrake-install.log && /bin/false)
+            --install femlium \
+            --package-branch tsfc JDBetteridge/constant_update \
+            --package-branch pyop2 JDBetteridge/global_literal \
+            --package-branch pyadjoint connorjward/constants-away || (cat firedrake-install.log && /bin/false)
```
Comment (dham) [inline .github/workflows/build.yml:80]:
```suggestion
            || (cat firedrake-install.log && /bin/false)
```
